crates/red_knot_python_semantic/src/types.rs

```diff
@@ -324,7 +326,9 @@ impl<'db> Type<'db> {
 
     pub const fn into_function_type(self) -> Option<FunctionType<'db>> {
         match self {
-            Type::Function(function_type) => Some(function_type),
+            Type::Function(function_type) | Type::RevealTypeFunction(function_type) => {
+                Some(function_type)
+            }
             _ => None,
         }
     }
```

Comment:
I guess the answer is probably *no* because the function is called `into_function_type` and not `into_function` but is there any existing code that makes the assumption that the function only returns `Some` if the variant is `Function`?

Reply:
I think this makes `Type::expect_function` not operate as expected, since the purpose of that function is to assert that the variant you have is a `Type::Function` variant:

https://github.com/astral-sh/ruff/blob/4eb849aed3a6df9ce0f45dd7a2fdb525e31e61e3/crates/red_knot_python_semantic/src/types.rs#L332-L335

Reply:
> is there any existing code that makes the assumption that the function only returns `Some` if the variant is `Function`?

No, there's no such code. And I think we generally want any external code to ideally transparently treat `Type::RevealTypeFunction` as if it were `Type::Function`. Because `reveal_type` is a function, and should behave as one.

This is another really good reason to store the inner `FunctionType` inside `RevealTypeFunction`.

> I think this makes `Type::expect_function` not operate as expected

I think this makes `Type::expect_function` behave precisely how we want it to, which is that the caller can transparently get out the correct `FunctionType`, and knows they are dealing with a function. Once the caller has a `FunctionType`, there's no reason they should care about the difference between `Type::Function` and `Type::RevealTypeFunction`.

Reply:
> I think this makes `Type::expect_function` behave precisely how we want it to

Well, you'll at least need to change the message in the `.expect()` call here, to something like "expected a variant that wraps a `FunctionType`"

https://github.com/astral-sh/ruff/blob/4eb849aed3a6df9ce0f45dd7a2fdb525e31e61e3/crates/red_knot_python_semantic/src/types.rs#L332-L335

Reply:
Good call! Fixed the message.